Refactor of `exists_locally` (and consequently a lot of other stuff...)

dafyddstephenson opened the pull request:
Branched from #74 , do not merge until after merging #74 

The initial purpose of this PR was a short refactor of the `exists_locally` attribute to allow the case when a yaml file describing an input dataset existed locally, but the associated netCDF file did not. It kept unspooling from there with no clear finish line. 

Ultimately, this is a lot of spring cleaning that is going to be necessary for the roms-tools integration. Changes to the user experience are fairly minimal (though have been updated in the example notebook), but what's happening behind the scenes is a lot cleaner and more intuitive, with hopefully all bodges and hacks removed.

closes #25 

## Summary:
- rename `local_path` attribute on `InputDataset` and `AdditionalCode` to `working_path` to avoid suggesting that the file(s) described by the path exist.
- Change `InputDataset.get(local_dir)` and `AdditionalCode.get(local_dir)` to actually fetch files to `local_dir`, not a convoluted subdirectory structure of it. `Case.setup()` now calls these methods with the required structure as an argument.
- Replace `check_exists_locally` on `InputDataset` and `AdditionalCode` with a property `exists_locally` which was not previously possible (and do the same for `check_is_setup` on `Case`, `check_local_config_status` on `BaseModel`)
- To make this possible, restructure of `AdditionalCode` in terms of expected directory/repo layout.
- ... Added an optional `subdir` attr/init argument to `AdditionalCode` in case the additional code lives in the subdirectory of a remote repository.
- Have changed the main example blueprint to remove MARBL additional code. Confusingly, as ROMS handles all input and output on MARBL's behalf, including reading things like namelist files, these are in fact _ROMS_ additional code. There were lots of hacky things in the code to get around this simple fact just because it didn't fit in very nicely with the C-Star design concept. They are now gone.
- The main example blueprint now points to [my fork](https://github.com/dafyddstephenson/roms_marbl_example/tree/main) of the example repo, which has been restructured, but I will update the C-Worthy version and C-Star blueprint once this PR is accepted. This was to avoid breaking tests for other people working with that repo.
- Provide explicit `output_dir` argument to `Component.run()` so it doesn't have to try and figure it out from the additional code directory structure anymore, which was ugly and confusing.

Commit:
begin rewriting repr functions: AdditionalCode, BaseModel, and Component

Commit:
add discretization class; some related tidying of Case.persist, Case.from_blueprint; no longer use kwargs: Any to initialise Component

Commit:
overhaul Case.from_blueprint to be cleaner, passing the dictionary read from yaml straight to the constructor method for each class and not repeating code for each kind of InputDataset

Commit:
tidy up and update docstrings

Commit:
update example notebook

Commit:
Merge branch 'main' into discretization_class

Commit:
fix bug where Component.discretization being None led to problems with Case.persist

Commit:
Merge branch 'discretization_class' into improve_reprs

Commit:
Merge branch 'pyproject_toml' into improve_reprs

Commit:
rewrite __str__ functions and add separate __repr__ functions

Commit:
add a tree() visualisation method for Case to get around the bash cell in the example notebook

Commit:
add discretization repr to component repr

dafyddstephenson committed:
Merge branch 'CWorthy-ocean:main' into improve_reprs

Commit:
- Replace InputDataset.local_path with InputDataset.working_path which refers to the path with which the dataset is being handled (e.g. in the context of a Case).
 The working_path does not strictly need to point to an existing file.
- Remove InputDataset.exists_locally attribute and InputDataset.check_exists_locally method.
  Existence of the file at working_path can be queried easily using pathlib (InputDataset.working_path.exists()).
- Update InputDataset.get(local_dir) to actually fetch the file to local_dir, not local_dir/input_datasets/<basemodelname>.
- Update Case.setup() to call InputDataset.get() with the correct directory path relative to the caseroot.
- Update docstrings and methods accordingly.

Commit:
Refactor AdditionalCode:
- As each AdditionalCode object is associated with a single BaseModel, remove references to BaseModel in the directory structure.
  This simplifies the directory structure to just additional_code_dir/namelists and additional_code_dir/source_mods
- Add a subdir attribute/argument to init. This is only really needed when 'location' is a remote git repo. It tells C-Star the
  additional code directory relative to the repo top level.
- Rename AdditionalCode.local_path to AdditionalCode.working_path, as was done for InputDataset in the previous commit
- Remove additional_code from MARBL section of the roms_marbl_example, and move MARBL namelists to ROMS namelists in that example.
  As the 'MARBL namelists' were actually read by ROMS, this was a misnomer and led to some weird bodges in the code.
- Accordingly, Make additional_code Optional in component.py (as it is no longer used by the MARBL Component)
- Update the ROMS MARBL example repo to point to my (restructured) fork for now until this PR is accepted.
- Make runtime-related paths in Component explicit rather than basing them on Component.additional_code.working_dir:
   - Add an 'output_dir' argument to ROMSComponent.run() and ROMSComponent.post_run() (and their calls in Case)

Commit:
linting

Commit:
More refactoring while testing:
- Move "check_property" fns and "is_property" boolean attrs into properties (in many cases was not previously possible):
   - Case.check_is_setup & Case.is_setup -> Case.is_setup
   - BaseModel.check_local_config_status & BaseModel.local_config_status -> BaseModel.local_config_status
   - AdditionalCode.check_exists_locally & AdditionalCode.exists_locally -> AdditionalCode.exists_locally
- Add "exists_locally" property back to InputDataset (just returns self.working_path.exists()) for consistency
- Add "is_setup" property to BaseModel (True if self.local_config_status is 0 else False) for consistency

Commit:
update docstrings, str and repr functions

dafyddstephenson committed:
Merge branch 'CWorthy-ocean:main' into abolish_exists_locally

Commit:
- Update ROMSComponent.post_run() to move raw output into PARTITIONED subdirectory after running ncjoin on it
- Update Case.tree() to reflect new AdditionalCode directory structure
- Update str function in AdditionalCode()
- Update example notebook to reflect all changes in this PR

TomNicholas commented on cstar/base/additional_code.py:
```diff
@@ -86,39 +94,31 @@ def __init__(
-            base_str += f"\n Local path: {self.local_path}"
+        base_str += f"\nsubdirectory: {self.subdir}"
+        base_str += f"\nWorking path: {self.working_path}"
+        base_str += f"\nExists locally: {self.exists_locally}"
```
You could use a multiline string instead of `+=` to add each line.

(`textwrap.dedent` will be useful)

dafyddstephenson replied:
I tried to keep each line modular to shift things around or add more things easily - though thinking about it a multi-line wouldn't really change that. Can do if it's more readable that way!

TomNicholas commented on cstar/base/additional_code.py:
```diff
@@ -86,39 +94,31 @@ def __init__(
+        base_str += f"\nWorking path: {self.working_path}"
+        base_str += f"\nExists locally: {self.exists_locally}"
+        if not self.exists_locally:
+            base_str += " (get with AdditionalCode.get())"
```
This might be a little more readable
```suggestion
            base_str.append(" (get with AdditionalCode.get())")
```

TomNicholas commented on cstar/base/component.py:
```diff
@@ -46,39 +55,33 @@ def __init__(self, **kwargs: Any):
-        base_str = "-" * (len(name) + 7) + "\n" + base_str
-        base_str += "\n" + "-" * (len(name) + 7)
+        base_str = f"{name}"
+        # base_str = "-" * len(name) + "\n" + base_str
```
```suggestion
```

TomNicholas commented on cstar/base/component.py:
```diff
@@ -87,43 +90,34 @@ def __str__(self) -> str:
+        if self.additional_code is not None:
+            repr_str += f"\nadditional_code = <{self.additional_code.__class__.__name__} instance>, "
+        else:
+            repr_str += "\n additional_code = None"
```
Isn't there some redundant here with the other classes? i.e. you could be calling `repr(additional_code)` here.

dafyddstephenson replied:
As the class structure is so hierarchical I tried to avoid a cascading printout of hundreds of lines by just calling reprs within reprs all the way down. This is a change from #74 however, would be good to discuss strs/reprs over there as that PR in particular was supposed to be a first stab at something that could become a discussion (the PR post itself also contains examples of all the strs/reprs so they're easier to read than just the code!)

TomNicholas reviewed (commented):
I think you've done a great job of making complex objects more understandable here :clap:

TomNicholas committed:
linting

TomNicholas committed:
fix mutable default type for input_datasets

Commit:
Tweaks following PR review:
- Split instances of multiple similar variables on one line in docstrings
- Remove nx,ny, levels attrs from Discretization class
- Reinstate type check for BaseModel in Component.__init__()

Commit:
update blueprint to reflect changes

Commit:
update example notebook

Commit:
Merge branch 'main' into discretization_class

NoraLoose committed:
Merge branch 'main' into discretization_class

Commit:
Merge branch 'discretization_class' into improve_reprs

dafyddstephenson committed:
Update cstar/base/component.py

Co-authored-by: Tom Nicholas <tom@cworthy.org>

dafyddstephenson committed:
Update cstar/base/component.py

Co-authored-by: Tom Nicholas <tom@cworthy.org>

Commit:
remove assert from component.py

TomNicholas committed:
Merge branch 'main' into abolish_exists_locally

Commit:
Merge branch 'main' into improve_reprs

Commit:
Merge branch 'improve_reprs' into abolish_exists_locally

Commit:
Merge branch 'main' into abolish_exists_locally

Commit:
Re-add default None for Component.additional_code (fell out in a merge somewhere)

dafyddstephenson commented:
ok some stuff got chewed up in the merges but I think we're good to go now, have updated the diffs here

TomNicholas commented on cstar/base/base_model.py:
```diff
@@ -294,3 +290,10 @@ def handle_config_status(self) -> None:
     @abstractmethod
     def get(self, target: str | Path) -> None:
```
This is not an urgent nor a particularly helpful comment at this time, but I think a less overloaded term for this method could be `fetch`, because `get` already has nother widely-used meaning in python.

dafyddstephenson replied:
Agree, I have been pondering a better alternative for a minute - even ignoring the conflict with python's use of get, ours is used in contexts going far beyond fetching, like `BaseModel.get()` which would be much better described as `BaseModel.setup()`. Fetch is fine, but in certain situations `get()` also creates symlinks, which I wouldn't really describe as "getting" or "fetching".

TomNicholas replied:
hmm, okay, let's punt on this. I'll raise your comment as an issue.

TomNicholas commented on cstar/base/base_model.py:
```diff
@@ -294,3 +290,10 @@ def handle_config_status(self) -> None:
+
+    @abstractmethod
+    def _base_model_adjustments(self) -> None:
+        """
+        Perform any C-Star specific adjustments to the base model that would
+        be needed after a clean checkout.
+        """
```
for example?

TomNicholas commented on cstar/base/input_dataset.py:
```diff
@@ -88,6 +75,14 @@ def __init__(
+    def exists_locally(self) -> bool:
+        return (
+            True
+            if ((self.working_path is not None) and (self.working_path.exists()))
```
Pretty sure most, if not all of these brackets are redundant
```suggestion
            if (self.working_path is not None) and self.working_path.exists()
```

dafyddstephenson replied:
They are, it's personal preference as most editors will highlight the matching bracket when you hover on one - makes it easier to see the nested structure / exactly what each condition is in chains of `and`s and `or`s

TomNicholas reviewed (approved):
Just some very minor comments.
